File: JsonSchemaAndStructs/KiteConnect-Typescript-Structs/gtt_get_order.ts
// To parse this data:
//
//   import { Convert, GttGetOrder } from "./file";
//
//   const gttGetOrder = Convert.toGttGetOrder(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

export interface GttGetOrder {
    data?:   Data;
    status?: string;
}

export interface Data {
    condition?:     Condition;
    createdAt?:     Date;
    expiresAt?:     Date;
    id?:            number;
    meta?:          null;
    orders?:        Order[];
    parentTrigger?: null;
    status?:        string;
    type?:          string;
    updatedAt?:     Date;
    userID?:        string;
}

export interface Condition {
    exchange?:        string;
    instrumentToken?: number;
    lastPrice?:       number;
    tradingsymbol?:   string;
    triggerValues?:   number[];
}

export interface Order {
    exchange?:        string;
    orderType?:       string;
    price?:           number;
    product?:         string;
    quantity?:        number;
    result?:          Result | null;
    tradingsymbol?:   string;
    transactionType?: string;
}

export interface Result {
    accountID?:       string;
    exchange?:        string;
    meta?:            string;
    orderResult?:     OrderResult;
    orderType?:       string;
    price?:           number;
    product?:         string;
    quantity?:        number;
    timestamp?:       Date;
    tradingsymbol?:   string;
    transactionType?: string;
    triggeredAt?:     number;
    validity?:        string;
}

export interface OrderResult {
    orderID?:         string;
    rejectionReason?: string;
    status?:          string;
}

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
export class Convert {
    public static toGttGetOrder(json: string): GttGetOrder {
        return cast(JSON.parse(json), r("GttGetOrder"));
    }

    public static gttGetOrderToJson(value: GttGetOrder): string {
        return JSON.stringify(uncast(value, r("GttGetOrder")), null, 2);
    }
}

function invalidValue(typ: any, val: any, key: any = ''): never {
    if (key) {
        throw Error(`Invalid value for key "${key}". Expected type ${JSON.stringify(typ)} but got ${JSON.stringify(val)}`);
    }
    throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`, );
}

function jsonToJSProps(typ: any): any {
    if (typ.jsonToJS === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ: any): any {
    if (typ.jsToJSON === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val: any, typ: any, getProps: any, key: any = ''): any {
    function transformPrimitive(typ: string, val: any): any {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val, key);
    }

    function transformUnion(typs: any[], val: any): any {
        // val must validate against one typ in typs
        const l = typs.length;
        for (let i = 0; i < l; i++) {
            const typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val);
    }

    function transformEnum(cases: string[], val: any): any {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases, val);
    }

    function transformArray(typ: any, val: any): any {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue("array", val);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformDate(val: any): any {
        if (val === null) {
            return null;
        }
        const d = new Date(val);
        if (isNaN(d.valueOf())) {
            return invalidValue("Date", val);
        }
        return d;
    }

    function transformObject(props: { [k: string]: any }, additional: any, val: any): any {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue("object", val);
        }
        const result: any = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps, prop.key);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps, key);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val);
    }
    if (typ === false) return invalidValue(typ, val);
    while (typeof typ === "object" && typ.ref !== undefined) {
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val);
    }
    // Numbers can be parsed by Date but shouldn't be.
    if (typ === Date && typeof val !== "number") return transformDate(val);
    return transformPrimitive(typ, val);
}

function cast<T>(val: any, typ: any): T {
    return transform(val, typ, jsonToJSProps);
}

function uncast<T>(val: T, typ: any): any {
    return transform(val, typ, jsToJSONProps);
}

function a(typ: any) {
    return { arrayItems: typ };
}

function u(...typs: any[]) {
    return { unionMembers: typs };
}

function o(props: any[], additional: any) {
    return { props, additional };
}

function m(additional: any) {
    return { props: [], additional };
}

function r(name: string) {
    return { ref: name };
}

const typeMap: any = {
    "GttGetOrder": o([
        { json: "data", js: "data", typ: u(undefined, r("Data")) },
        { json: "status", js: "status", typ: u(undefined, "") },
    ], false),
    "Data": o([
        { json: "condition", js: "condition", typ: u(undefined, r("Condition")) },
        { json: "created_at", js: "createdAt", typ: u(undefined, Date) },
        { json: "expires_at", js: "expiresAt", typ: u(undefined, Date) },
        { json: "id", js: "id", typ: u(undefined, 0) },
        { json: "meta", js: "meta", typ: u(undefined, null) },
        { json: "orders", js: "orders", typ: u(undefined, a(r("Order"))) },
        { json: "parent_trigger", js: "parentTrigger", typ: u(undefined, null) },
        { json: "status", js: "status", typ: u(undefined, "") },
        { json: "type", js: "type", typ: u(undefined, "") },
        { json: "updated_at", js: "updatedAt", typ: u(undefined, Date) },
        { json: "user_id", js: "userID", typ: u(undefined, "") },
    ], false),
    "Condition": o([
        { json: "exchange", js: "exchange", typ: u(undefined, "") },
        { json: "instrument_token", js: "instrumentToken", typ: u(undefined, 0) },
        { json: "last_price", js: "lastPrice", typ: u(undefined, 3.14) },
        { json: "tradingsymbol", js: "tradingsymbol", typ: u(undefined, "") },
        { json: "trigger_values", js: "triggerValues", typ: u(undefined, a(3.14)) },
    ], false),
    "Order": o([
        { json: "exchange", js: "exchange", typ: u(undefined, "") },
        { json: "order_type", js: "orderType", typ: u(undefined, "") },
        { json: "price", js: "price", typ: u(undefined, 0) },
        { json: "product", js: "product", typ: u(undefined, "") },
        { json: "quantity", js: "quantity", typ: u(undefined, 0) },
        { json: "result", js: "result", typ: u(undefined, u(r("Result"), null)) },
        { json: "tradingsymbol", js: "tradingsymbol", typ: u(undefined, "") },
        { json: "transaction_type", js: "transactionType", typ: u(undefined, "") },
    ], false),
    "Result": o([
        { json: "account_id", js: "accountID", typ: u(undefined, "") },
        { json: "exchange", js: "exchange", typ: u(undefined, "") },
        { json: "meta", js: "meta", typ: u(undefined, "") },
        { json: "order_result", js: "orderResult", typ: u(undefined, r("OrderResult")) },
        { json: "order_type", js: "orderType", typ: u(undefined, "") },
        { json: "price", js: "price", typ: u(undefined, 0) },
        { json: "product", js: "product", typ: u(undefined, "") },
        { json: "quantity", js: "quantity", typ: u(undefined, 0) },
        { json: "timestamp", js: "timestamp", typ: u(undefined, Date) },
        { json: "tradingsymbol", js: "tradingsymbol", typ: u(undefined, "") },
        { json: "transaction_type", js: "transactionType", typ: u(undefined, "") },
        { json: "triggered_at", js: "triggeredAt", typ: u(undefined, 3.14) },
        { json: "validity", js: "validity", typ: u(undefined, "") },
    ], false),
    "OrderResult": o([
        { json: "order_id", js: "orderID", typ: u(undefined, "") },
        { json: "rejection_reason", js: "rejectionReason", typ: u(undefined, "") },
        { json: "status", js: "status", typ: u(undefined, "") },
    ], false),
};

File: JsonSchemaAndStructs/KiteConnect-Python-DataClass/gtt_get_order.py
# This code parses date/times, so please
#
#     pip install python-dateutil
#
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = gtt_get_order_from_dict(json.loads(json_string))

from dataclasses import dataclass
from typing import Optional, List, Any, TypeVar, Callable, Type, cast
from datetime import datetime
import dateutil.parser


T = TypeVar("T")


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def from_none(x: Any) -> Any:
    assert x is None
    return x


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def from_float(x: Any) -> float:
    assert isinstance(x, (float, int)) and not isinstance(x, bool)
    return float(x)


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def to_float(x: Any) -> float:
    assert isinstance(x, float)
    return x


def from_datetime(x: Any) -> datetime:
    return dateutil.parser.parse(x)


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


@dataclass(slots=True)
class Condition:
    exchange: Optional[str] = None
    instrument_token: Optional[int] = None
    last_price: Optional[float] = None
    tradingsymbol: Optional[str] = None
    trigger_values: Optional[List[float]] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Condition':
        assert isinstance(obj, dict)
        exchange = from_union([from_str, from_none], obj.get("exchange"))
        instrument_token = from_union([from_int, from_none], obj.get("instrument_token"))
        last_price = from_union([from_float, from_none], obj.get("last_price"))
        tradingsymbol = from_union([from_str, from_none], obj.get("tradingsymbol"))
        trigger_values = from_union([lambda x: from_list(from_float, x), from_none], obj.get("trigger_values"))
        return Condition(exchange, instrument_token, last_price, tradingsymbol, trigger_values)

    def to_dict(self) -> dict:
        result: dict = {}
        result["exchange"] = from_union([from_str, from_none], self.exchange)
        result["instrument_token"] = from_union([from_int, from_none], self.instrument_token)
        result["last_price"] = from_union([to_float, from_none], self.last_price)
        result["tradingsymbol"] = from_union([from_str, from_none], self.tradingsymbol)
        result["trigger_values"] = from_union([lambda x: from_list(to_float, x), from_none], self.trigger_values)
        return result


@dataclass(slots=True)
class OrderResult:
    order_id: Optional[str] = None
    rejection_reason: Optional[str] = None
    status: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'OrderResult':
        assert isinstance(obj, dict)
        order_id = from_union([from_str, from_none], obj.get("order_id"))
        rejection_reason = from_union([from_str, from_none], obj.get("rejection_reason"))
        status = from_union([from_str, from_none], obj.get("status"))
        return OrderResult(order_id, rejection_reason, status)

    def to_dict(self) -> dict:
        result: dict = {}
        result["order_id"] = from_union([from_str, from_none], self.order_id)
        result["rejection_reason"] = from_union([from_str, from_none], self.rejection_reason)
        result["status"] = from_union([from_str, from_none], self.status)
        return result


@dataclass(slots=True)
class Result:
    account_id: Optional[str] = None
    exchange: Optional[str] = None
    meta: Optional[str] = None
    order_result: Optional[OrderResult] = None
    order_type: Optional[str] = None
    price: Optional[int] = None
    product: Optional[str] = None
    quantity: Optional[int] = None
    timestamp: Optional[datetime] = None
    tradingsymbol: Optional[str] = None
    transaction_type: Optional[str] = None
    triggered_at: Optional[float] = None
    validity: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Result':
        assert isinstance(obj, dict)
        account_id = from_union([from_str, from_none], obj.get("account_id"))
        exchange = from_union([from_str, from_none], obj.get("exchange"))
        meta = from_union([from_str, from_none], obj.get("meta"))
        order_result = from_union([OrderResult.from_dict, from_none], obj.get("order_result"))
        order_type = from_union([from_str, from_none], obj.get("order_type"))
        price = from_union([from_int, from_none], obj.get("price"))
        product = from_union([from_str, from_none], obj.get("product"))
        quantity = from_union([from_int, from_none], obj.get("quantity"))
        timestamp = from_union([from_datetime, from_none], obj.get("timestamp"))
        tradingsymbol = from_union([from_str, from_none], obj.get("tradingsymbol"))
        transaction_type = from_union([from_str, from_none], obj.get("transaction_type"))
        triggered_at = from_union([from_float, from_none], obj.get("triggered_at"))
        validity = from_union([from_str, from_none], obj.get("validity"))
        return Result(account_id, exchange, meta, order_result, order_type, price, product, quantity, timestamp, tradingsymbol, transaction_type, triggered_at, validity)

    def to_dict(self) -> dict:
        result: dict = {}
        result["account_id"] = from_union([from_str, from_none], self.account_id)
        result["exchange"] = from_union([from_str, from_none], self.exchange)
        result["meta"] = from_union([from_str, from_none], self.meta)
        result["order_result"] = from_union([lambda x: to_class(OrderResult, x), from_none], self.order_result)
        result["order_type"] = from_union([from_str, from_none], self.order_type)
        result["price"] = from_union([from_int, from_none], self.price)
        result["product"] = from_union([from_str, from_none], self.product)
        result["quantity"] = from_union([from_int, from_none], self.quantity)
        result["timestamp"] = from_union([lambda x: x.isoformat(), from_none], self.timestamp)
        result["tradingsymbol"] = from_union([from_str, from_none], self.tradingsymbol)
        result["transaction_type"] = from_union([from_str, from_none], self.transaction_type)
        result["triggered_at"] = from_union([to_float, from_none], self.triggered_at)
        result["validity"] = from_union([from_str, from_none], self.validity)
        return result


@dataclass(slots=True)
class Order:
    exchange: Optional[str] = None
    order_type: Optional[str] = None
    price: Optional[int] = None
    product: Optional[str] = None
    quantity: Optional[int] = None
    result: Optional[Result] = None
    tradingsymbol: Optional[str] = None
    transaction_type: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Order':
        assert isinstance(obj, dict)
        exchange = from_union([from_str, from_none], obj.get("exchange"))
        order_type = from_union([from_str, from_none], obj.get("order_type"))
        price = from_union([from_int, from_none], obj.get("price"))
        product = from_union([from_str, from_none], obj.get("product"))
        quantity = from_union([from_int, from_none], obj.get("quantity"))
        result = from_union([Result.from_dict, from_none], obj.get("result"))
        tradingsymbol = from_union([from_str, from_none], obj.get("tradingsymbol"))
        transaction_type = from_union([from_str, from_none], obj.get("transaction_type"))
        return Order(exchange, order_type, price, product, quantity, result, tradingsymbol, transaction_type)

    def to_dict(self) -> dict:
        result: dict = {}
        result["exchange"] = from_union([from_str, from_none], self.exchange)
        result["order_type"] = from_union([from_str, from_none], self.order_type)
        result["price"] = from_union([from_int, from_none], self.price)
        result["product"] = from_union([from_str, from_none], self.product)
        result["quantity"] = from_union([from_int, from_none], self.quantity)
        result["result"] = from_union([lambda x: to_class(Result, x), from_none], self.result)
        result["tradingsymbol"] = from_union([from_str, from_none], self.tradingsymbol)
        result["transaction_type"] = from_union([from_str, from_none], self.transaction_type)
        return result


@dataclass(slots=True)
class Data:
    meta: None
    parent_trigger: None
    condition: Optional[Condition] = None
    created_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    id: Optional[int] = None
    orders: Optional[List[Order]] = None
    status: Optional[str] = None
    type: Optional[str] = None
    updated_at: Optional[datetime] = None
    user_id: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Data':
        assert isinstance(obj, dict)
        meta = from_none(obj.get("meta"))
        parent_trigger = from_none(obj.get("parent_trigger"))
        condition = from_union([Condition.from_dict, from_none], obj.get("condition"))
        created_at = from_union([from_datetime, from_none], obj.get("created_at"))
        expires_at = from_union([from_datetime, from_none], obj.get("expires_at"))
        id = from_union([from_int, from_none], obj.get("id"))
        orders = from_union([lambda x: from_list(Order.from_dict, x), from_none], obj.get("orders"))
        status = from_union([from_str, from_none], obj.get("status"))
        type = from_union([from_str, from_none], obj.get("type"))
        updated_at = from_union([from_datetime, from_none], obj.get("updated_at"))
        user_id = from_union([from_str, from_none], obj.get("user_id"))
        return Data(meta, parent_trigger, condition, created_at, expires_at, id, orders, status, type, updated_at, user_id)

    def to_dict(self) -> dict:
        result: dict = {}
        result["meta"] = from_none(self.meta)
        result["parent_trigger"] = from_none(self.parent_trigger)
        result["condition"] = from_union([lambda x: to_class(Condition, x), from_none], self.condition)
        result["created_at"] = from_union([lambda x: x.isoformat(), from_none], self.created_at)
        result["expires_at"] = from_union([lambda x: x.isoformat(), from_none], self.expires_at)
        result["id"] = from_union([from_int, from_none], self.id)
        result["orders"] = from_union([lambda x: from_list(lambda x: to_class(Order, x), x), from_none], self.orders)
        result["status"] = from_union([from_str, from_none], self.status)
        result["type"] = from_union([from_str, from_none], self.type)
        result["updated_at"] = from_union([lambda x: x.isoformat(), from_none], self.updated_at)
        result["user_id"] = from_union([from_str, from_none], self.user_id)
        return result


@dataclass(slots=True)
class GttGetOrder:
    data: Optional[Data] = None
    status: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'GttGetOrder':
        assert isinstance(obj, dict)
        data = from_union([Data.from_dict, from_none], obj.get("data"))
        status = from_union([from_str, from_none], obj.get("status"))
        return GttGetOrder(data, status)

    def to_dict(self) -> dict:
        result: dict = {}
        result["data"] = from_union([lambda x: to_class(Data, x), from_none], self.data)
        result["status"] = from_union([from_str, from_none], self.status)
        return result


def gtt_get_order_from_dict(s: Any) -> GttGetOrder:
    return GttGetOrder.from_dict(s)


def gtt_get_order_to_dict(x: GttGetOrder) -> Any:
    return to_class(GttGetOrder, x)
